src/cli/commands/block/put.js

```diff
@@ -1,52 +1,92 @@
 'use strict'
 
 const utils = require('../../utils')
-const mh = require('multihashes')
 const bl = require('bl')
 const fs = require('fs')
 const Block = require('ipfs-block')
+const CID = require('cids')
 const waterfall = require('async/waterfall')
 const debug = require('debug')
 const log = debug('cli:block')
 log.error = debug('cli:block:error')
 
-function addBlock (buf) {
+function addBlock (buf, opts) {
+  let block = new Block(buf)
+  let mhash, cid
+
   utils.getIPFS((err, ipfs) => {
     if (err) {
       throw err
     }
 
     waterfall([
-      (cb) => ipfs.block.put(new Block(buf), cb),
-      (block, cb) => block.key(cb)
-    ], (err, key) => {
+      (cb) => generateHash(block, opts, cb),
+      (cb) => generateCid(mhash, block, opts, cb),
+      (cb) => ipfs.block.put(block, cid, cb)
+    ], (err) => {
       if (err) {
         throw err
       }
 
-      console.log(mh.toB58String(key))
+      console.log(cid.toBaseEncodedString())
     })
   })
+
+  function generateHash (block, opts, cb) {
+    if (opts.format === 'v0') {
+      block.key(done)
+    } else {
+      block.key(opts.mhtype, done)
+    }
+    function done (err, _mhash) {
+      if (err) return cb(err)
+      mhash = _mhash
+      cb()
+    }
```

Comment:
since you are doing waterfall, you can pass the `cb` directly to the block.key function and capture the hash in the waterfall sequence